### ISSUE TITLE
setsockopt(TCP_NODELAY) in Windows

### ISSUE THREAD

**Issue by michael-grunder** [redis/hiredis#785]
There appears to be a Windows-specific issue setting TCP_NODELAY in non-blocking mode.

See related issues #736, #779.

Yielding the timeslice (with `_sleep(0)`) solves the problem, but I have not been able to verify if it is the *correct* way to solve the problem.

If it is correct, perhaps adding it to win32_setsockopt is the cleanest solution:

```c
diff --git sockcompat.c sockcompat.c
index f99d14b..b7990b8 100644
--- sockcompat.c
+++ sockcompat.c
@@ -216,6 +216,10 @@ int win32_setsockopt(SOCKET sockfd, int level, int optname, const void *optval,
         DWORD timeout = tv->tv_sec * 1000 + tv->tv_usec / 1000;
         ret = setsockopt(sockfd, level, optname, (const char*)&timeout, sizeof(DWORD));
     } else {
+#if _MSC_VER
+        /* Windows workaround */
+        if (optname == TCP_NODELAY) _sleep(0);
+#endif
         ret = setsockopt(sockfd, level, optname, (const char*)optval, optlen);
     }
     _updateErrno(ret != SOCKET_ERROR);
```

@mbitsnbites Do you mind weighing in as you wrote the windows compatibility layer?

Also curious if @mnunberg or @yossigo have opinions one way or the other.

**Comment by mbitsnbites** [comment]
I have no real opinion regarding the fix (it's above my winsock competence level). But I don't get the `#if _MSC_VER` part. Does the problem only appear in Visual Studio builds? All code in the `sockcompat.c` file is Windows specific, so if the problem is a Windows problem, no ifdef should be needed.

**Comment by michael-grunder** [comment]
> Does the problem only appear in Visual Studio builds

I honestly don't know but it's a good point so I will try to verify.

Back to digging around in Windows TCP code to see if I can find a definitive answer.

Thanks for taking a look though!

**Comment by michael-grunder** [comment]
@xingyun86 Is the `_sleep(0)` call only needed on older versions of Windows?  I've been doing a bit of testing and haven't encountered any errors setting TCP_NODELAY in a Win10 VM.

I want to get the fix in I'm just nervous about doing that without understanding _why_ it is needed.  :smile:

**Comment by michael-grunder** [comment]
I did a bit of digging on this and think I know what's happening and how we might be able to solve it.

In `sockcompat.c` we have the utility function `_wsaErrorToErrorno` that converts `WSAGetLastError` codes to their Linux equivelants.  When it encounters an error that isn't in the switch it defaults to [EIO](https://github.com/redis/hiredis/blob/0184caac9de25258b23156016b3528e9878f1dae/sockcompat.c#L93), which I'm guessing is how you end up seeing `Input/Output error`.

Here's a [list of the uncovered errors](https://gist.github.com/michael-grunder/57a274337fad5d8ea7f0efcae2339a6c).

I haven't been able to make the `setsockopt` call fail, but if someone can we might be able to solve this problem in the _right_ way.

@xingyun86 If you are able to cause the error and post the exact WSA error number it would be awesome!

**Comment by yossigo** [comment]
@michael-grunder It will be a good idea to make sure all errors are handled by `_wsaErrorToErrorno` properly, but I can't see how it will address this issue.

If that's indeed a Windows issue with calling `setsockopt()` on a socket in some intermediate connection state, the right way to solve that may be different for async and blocking connections. While it may be OK for a blocking connect to have an extra check to be sure the socket is indeed ready, async connects would have to just postpone that and do it in the connect callback.

Anyway I agree it needs to be reproduced before applying any kind of fix.

**Comment by michael-grunder** [comment]
>  It will be a good idea to make sure all errors are handled by _wsaErrorToErrorno properly, but I can't see how it will address this issue

What I meant is that perhaps the actual error coming back from `WSAGetLastError` is something akin to `EAGAIN` (maybe `WSA_IO_INCOMPLETE`, or `WSA_IO_PENDING`)?

In the connect logic we do something like:

1.  redisSetBlocking(c, 0);
2.  a bunch of connection stuff
3.  redisSetBlocking(c, 1);
4.  redisSetTcpNodelay(c);

Perhaps immediately setting `TCP_NODELAY` after setting the socket back to blocking is failing with some kind of "not ready yet" error.  This could be right since `_sleep(0)` fixes it.

This is just wild speculation though.  I haven't developed on windows since the 90s :smile:

**Comment by yossigo** [comment]
@michael-grunder Oh right! I didn't notice plausible error codes are not handled.

**Comment by amitv87** [comment]
I use mac and get the error ```setsockopt(TCP_NODELAY): Invalid argument``` intermittently when **redisAsyncConnect** is called. Don't think if this is windows only issue

**Comment by michael-grunder** [comment]
That is interesting.  I've never seen that happen on OSX but of course, that doesn't mean it's not a real bug.

An `Invalid argument` error would mean we're either setting a nonsensical value (I suppose something other than `0` or `1` in the case of TCP_NODELAY) or the socket has already been shut down.

**Comment by chris-shenzhen-china** [comment]
I set socket keepalive option first, then call set TCP_NODELAY  option, it works.

static int redisSetTcpNoDelay(redisContext *c) {
    int ret = 0;
    ret = setSockKeepAlive(c->fd, 1);
    if (0 != ret)
    {
        redisNetClose(c);
        return REDIS_ERR;
    }

    int yes = 1;
    if (setsockopt(c->fd, IPPROTO_TCP, TCP_NODELAY, &yes, sizeof(yes)) == -1) {
        __redisSetErrorFromErrno(c,REDIS_ERR_IO,"setsockopt(TCP_NODELAY)");
        redisNetClose(c);
        return REDIS_ERR;
    }
    return REDIS_OK;
}

/* set keepalive socket option */
int setSockKeepAlive(SOCKET sockfd, int val)
{
    int   keepalive;
    int ret = 0;
    keepalive = (int)val;

    ret = setsockopt( sockfd, SOL_SOCKET, SO_KEEPALIVE, (char *)&keepalive, sizeof(keepalive) );
    _updateErrno(ret != SOCKET_ERROR);
    return ret != SOCKET_ERROR ? ret : -1;
}

**Comment by hunterfortuin** [comment]
I'm seeing `WSAEINVAL` come through on Windows 10.

**Comment by OmriSteiner** [comment]
Hey, I've experienced this bug as well, and am able to reproduce it consistently.
It seems like `setsockopt` returns `10022`, aka `WSAEINVAL`.

It seems like a race condition. I am guessing the `setsockopt` fails if it is called before the connection completed successfully.

I think the best solution is, like @yossigo said, to defer the `setsockopt` call in async connections. I've created a pull request which implements this.
https://github.com/redis/hiredis/pull/836
